apps/arweave/src/ar_mining_worker.erl

```diff
@@ -541,53 +473,65 @@ handle_task({computed_h2, Candidate, _ExtraArgs}, State) ->
-					PoA;
+				{ok, PoA} -> PoA;
 				{error, _Error} ->
 					%% Fallback. This will probably fail later, but prepare_poa/3 should
 					%% have already printed several errors so we'll continue just in case.
+					%% df: Is this the right fallback?..
 					#poa{ chunk = Chunk2 }
 			end,
 			ar_coordination:computed_h2_for_peer(Candidate#mining_candidate{ poa2 = PoA2 })
 	end,
-	{noreply, State2};
+	%% Remove the cached value from the cache.
+	case ar_mining_cache:with_cached_value(
+		?VDF_STEP_CACHE_KEY(Candidate#mining_candidate.cache_ref, Candidate#mining_candidate.nonce),
+		Candidate#mining_candidate.session_key,
+		State2#state.chunk_cache,
+		fun(_) -> {ok, drop} end
+	) of
+		{ok, ChunkCache2} -> {noreply, State2#state{ chunk_cache = ChunkCache2 }};
+		{error, Reason} ->
+			?LOG_ERROR([{event, mining_worker_failed_to_process_computed_h2},
+				{worker, State2#state.name}, {partition, State2#state.partition_number},
+				{nonce, Candidate#mining_candidate.nonce},
+				{session_key, ar_nonce_limiter:encode_session_key(Candidate#mining_candidate.session_key)},
+				{reason, Reason}]),
+			{noreply, State2}
+	end;
 
+%% @doc Handle the `compute_h2_for_peer` task.
+%% Indicates that we got a request to compute h2 for a peer.
 handle_task({compute_h2_for_peer, Candidate, _ExtraArgs}, State) ->
 	#mining_candidate{
-		session_key = SessionKey,
 		h0 = H0,
 		partition_number = Partition1,
 		partition_upper_bound = PartitionUpperBound,
 		cm_h1_list = H1List,
-		cm_lead_peer = Peer,
-		packing_difficulty = PackingDifficulty
+		cm_lead_peer = Peer
 	} = Candidate,
-
-	{_RecallRange1Start, RecallRange2Start} = ar_block:get_recall_range(H0,
-					Partition1, PartitionUpperBound),
+	{_, RecallRange2Start} = ar_block:get_recall_range(H0, Partition1, PartitionUpperBound),
 	Candidate2 = generate_cache_ref(Candidate),
-	%% Clear the list so we aren't copying it around all over the place
+	%% Clear the list so we aren't copying it around all over the place.
 	Candidate3 = Candidate2#mining_candidate{ cm_h1_list = [] },
-
 	Range2Exists = ar_mining_io:read_recall_range(chunk2, self(), Candidate3, RecallRange2Start),
 	case Range2Exists of
 		true ->
 			ar_mining_stats:h1_received_from_peer(Peer, length(H1List)),
-			%% Note: when processing CM requests we always reserve the cache space and proceed
-			%% *even if* this puts us over the chunk cache limit. This may have to be
-			%% revisited later if we find that this causes unacceptable memory bloat.
-			State2 = update_sub_chunk_cache_size(
-				ar_block:get_nonces_per_recall_range(PackingDifficulty), SessionKey, State),
-			%% First flag all nonces in the range as do_not_cache, then cache the specific
-			%% nonces included in the H1 list. This will make sure we don't cache the chunk2s
-			%% that are read for the missing nonces.
-			State3 = do_not_cache(Candidate3, State2),
-			{noreply, cache_h1_list(Candidate3, H1List, State3)};
+			State2 = mark_chunk2_missing(Candidate3, State),
```

Comment:
is marking chunk2 missing just a sanity check? It should be missing by default right?

Reply:
Ah, it was `do_not_cache` before. It was used not to cache the `chunk2`s that we will not use, only those that correspond to received `h1`s. It is now broken I think, I will need to handle this case in `chunk2` event.